### PR TITLE
Handle APIC EOI message for MSHV

### PR DESCRIPTION
**Pull request by liuw**
It is discovered in our testing that MSHV didn't handle this message.

In general I don't think Cloud Hypervisor handles level-triggered interrupts well. KVM code is missing resamplefd code. That's something to be fixed later.

### CODE REVIEW

**Review by jinankjain — APPROVED** [review]

File: hypervisor/src/mshv/mod.rs [hunk 1]
```diff
@@ -562,6 +562,18 @@ impl cpu::Vcpu for MshvVcpu {
                     debug!("Exception Info {:?}", { info.exception_vector });
                     Ok(cpu::VmExit::Ignore)
                 }
+                hv_message_type_HVMSG_X64_APIC_EOI => {
```
Comment (russell-islam) [inline hypervisor/src/mshv/mod.rs:565]: Anyway we get an exit here? I think we are not.

Comment (praveen-pk) [reply hypervisor/src/mshv/mod.rs:565]: During our testing we encountered this exit here while passing through a PCI device to guest, disabling MSIX support.

File: hypervisor/src/mshv/mod.rs [hunk 2]
```diff
@@ -562,6 +562,18 @@ impl cpu::Vcpu for MshvVcpu {
...
+                    let info = x.to_apic_eoi_info().unwrap();
+                    // The kernel should dispatch the EOI to the correct thread.
+                    // Check the VP index is the same as the one we have.
+                    assert!(info.vp_index == self.vp_index as u32);
```
Comment (praveen-pk) [inline hypervisor/src/mshv/mod.rs:569]: Is it possible for EOI to be delivered to a different VP (than the one where the interrupt was received)? If such a thing does happen, does it mean the hypervisor is faulty?

Comment (liuw) [reply hypervisor/src/mshv/mod.rs:569]: Any VP running inside the guest can generate such an event. Hypervisor doesn't really shuffle things from one VP to another. Like I said in the comment, if the VP index is wrong then there is a bug in our kernel driver.

Comment (praveen-pk) [reply hypervisor/src/mshv/mod.rs:569]: Thanks for clarifying.

File: hypervisor/src/mshv/mod.rs (hunk 1 above)
Comment (liuw) [reply hypervisor/src/mshv/mod.rs:565]: I think we can only get this if we enable legacy PCI INTx support. It uses level-triggered interrupt, which requires explicitly acking the vector.

**Review by praveen-pk — APPROVED** [review]

**Review by russell-islam — APPROVED** [review]

File: hypervisor/src/mshv/mod.rs (hunk 1 above)
Comment (rbradford) [reply hypervisor/src/mshv/mod.rs:565]: We have PCI INTx support for VFIO and vfio-user devices.

**Review by rbradford — APPROVED** [review]
You can merge this is if you like but I see lots of issues about EOI and resamplefd in the issue tracker.

**Pull request merged**